alg/teca_detect_nodes.cxx

```diff
@@ -0,0 +1,1945 @@
+          {
+             for (int jx = -1; jx <= 1; jx++)
+             {
+                if ((ix == 0) && (jx == 0))
+                {
+                   continue;
+                }
+
+                int inew = i + ix;
+                int jnew = j + jx;
+
+                if ((jnew < 0) || (jnew >= nLat))
+                {
+                   neighbors++;
+                   continue;
+                }
+                if (inew < 0)
+                {
+                   inew += nLon;
+                }
+                if (inew >= nLon)
+                {
+                   inew -= nLon;
+                }
+
+                p_vecConnectivity[neighbors * nLat * nLon + q] = jnew * nLon + inew;
+                neighbors++;
+             }
+          }
+       }
+
+    // Connectivity in the four primary directions
+    } else {
+       if (j != 0)
+       {
+          p_vecConnectivity[neighbors * nLat * nLon + q] = (j-1) * nLon + i;
+          neighbors++;
+       }
+
+       if (j != nLat-1)
+       {
+          p_vecConnectivity[neighbors * nLat * nLon + q] = (j+1) * nLon + i;
+          neighbors++;
+       }
+
+       if ((!fRegional) || ((i != 0) && (i != nLon-1)))
+       {
+          p_vecConnectivity[neighbors * nLat * nLon + q] = j * nLon + ((i + 1) % nLon);
+          neighbors++;
+          p_vecConnectivity[neighbors * nLat * nLon + q] = j * nLon + ((i + nLon - 1) % nLon);
+       }
+       //if (fRegional)
+       //{
+       //   if (i != 0)
+       //   {
+       //      p_vecConnectivity[neighbors * nLat * nLon + q] = q - 1;
+       //   }
+       //   neighbors++;
+
+       //   if (i != nLon-1)
+       //   {
+       //      p_vecConnectivity[neighbors * nLat * nLon + q] = q + 1;
+       //   }
+       //} else {
+       //   p_vecConnectivity[neighbors * nLat * nLon + q] = j * nLon + ((i + nLon - 1) % nLon);
+       //   neighbors++;
+       //   p_vecConnectivity[neighbors * nLat * nLon + q] = j * nLon + ((i + 1) % nLon);
+       //}
+    }
+}
+// --------------------------------------------------------------------------
+template <typename T, typename T_SEARCH>
+__global__
+void find_all_local_minima(
+     T *p_vecConnectivity,
+     const T_SEARCH *p_data,
+     T_SEARCH *p_setMinima,
+     int sNeighbors,
+     unsigned long nLat,
+     unsigned long nLon)
+{
+    unsigned long q = teca_cuda_util::thread_id_to_array_index();
+
+    if (q >= nLat*nLon)
+       return;
+
+    bool fMinimum = true;
+
+    for (int n = 0; n < sNeighbors; n++)
+    {
+       if (p_vecConnectivity[n * nLat * nLon + q] >= 0)
+       {
+          if (p_data[int(p_vecConnectivity[n * nLat * nLon + q])] < p_data[q])
+          {
+             fMinimum = false;
+             break;
+          }
+       }
+    }
+
+    if (fMinimum)
+    {
+       p_setMinima[q] = q;
+    }
+}
+// --------------------------------------------------------------------------
+template <typename T, typename T_SEARCH>
+__global__
+void find_all_local_maxima(
```

Comment:
why are there two versions of this function (and of the _threshold variant of this function)?  Based on the way the're called later on the code, I'm thinking one is a GPU version and the other is a CPU version: is that right?  It might be preferable to update the names so that this distinction is more clear if so: e.g., `find_all_local_maxima_gpu()`

Reply:
Done!